calibrate-touchscreen/src/main.rs

```diff
@@ -0,0 +1,429 @@
+                self.touches[1].1*(self.touches[0].0*TARGETS[2].1 - self.touches[2].0*TARGETS[0].1) +
+                self.touches[2].1*(self.touches[1].0*TARGETS[0].1 - self.touches[0].0*TARGETS[1].1);
+
+            println!("Calibration = {:.3}, {:.3}, {:.3}, {:.3}, {:.3}, {:.3}", ak/k, bk/k, ck/k, dk/k, ek/k, fk/k);
+
+            self.running = false;
+        }
+    }
+
+    fn size_of(&self, surface: &wl_surface::WlSurface) -> (f64, f64) {
+        for (_,fs) in &self.outputs {
+            let ss = fs.as_ref().unwrap();
+            if ss.wl_surface.as_ref().unwrap() == surface {
+                return (ss.width as f64, ss.height as f64);
+            };
+        }
+        unreachable!("The loop should always return");
+    }
+}
+
+impl Dispatch<xdg_wm_base::XdgWmBase, ()> for State {
+    fn event(
+        _: &mut Self,
+        wm_base: &xdg_wm_base::XdgWmBase,
+        event: xdg_wm_base::Event,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<Self>,
+    ) {
+        if let xdg_wm_base::Event::Ping { serial } = event {
+            wm_base.pong(serial);
+        }
+    }
+}
+
+impl Dispatch<xdg_surface::XdgSurface, ()> for State {
+    fn event(
+        _: &mut Self,
+        _: &xdg_surface::XdgSurface,
+        _: xdg_surface::Event,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<Self>,
+    ) {
+    }
+}
+
+impl Dispatch<xdg_toplevel::XdgToplevel, ()> for State {
+    fn event(
+        state: &mut Self,
+        _: &xdg_toplevel::XdgToplevel,
+        event: xdg_toplevel::Event,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<Self>,
+    ) {
+        if let xdg_toplevel::Event::Close {} = event {
+            state.running = false;
+        }
+    }
+}
+
+impl Dispatch<wl_seat::WlSeat, ()> for State {
+    fn event(
+        _: &mut Self,
+        seat: &wl_seat::WlSeat,
+        event: wl_seat::Event,
+        _: &(),
+        _: &Connection,
+        qh: &QueueHandle<Self>,
+    ) {
+        match event {
+            wl_seat::Event::Capabilities { capabilities: WEnum::Value(capabilities) } => {
+                if capabilities.contains(wl_seat::Capability::Keyboard) {
+                    seat.get_keyboard(qh, ());
+                }
+                if capabilities.contains(wl_seat::Capability::Pointer) {
+                    seat.get_pointer(qh, ());
+                }
+                if capabilities.contains(wl_seat::Capability::Touch) {
+                    seat.get_touch(qh, ());
+                }
+            }
+            _ => {}
+        }
+    }
+}
+
+impl Dispatch<wl_keyboard::WlKeyboard, ()> for State {
+    fn event(
+        state: &mut Self,
+        _: &wl_keyboard::WlKeyboard,
+        event: wl_keyboard::Event,
+        _: &(),
+        _: &Connection,
+        _: &QueueHandle<Self>,
+    ) {
+        if let wl_keyboard::Event::Key { key, .. } = event {
+            if key == 1 {
```

Comment:
I'm *fairly* sure this can be simplified to
```suggestion
        if  wl_keyboard::Event::Key { 1, .. } = event {
```

Reply:
Or something nearly there? Definitely this could be a `match`.